File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/test/ExponentTestNativeModule.java
// Copyright 2015-present 650 Industries. All rights reserved.

package versioned.host.exp.exponent.modules.test;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;

import de.greenrobot.event.EventBus;
import host.exp.exponent.test.TestCompletedEvent;
import host.exp.exponent.test.TestActionEvent;

public class ExponentTestNativeModule extends ReactContextBaseJavaModule {

  public ExponentTestNativeModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  public String getName() {
    return "ExponentTest";
  }

  @ReactMethod
  public void action(final ReadableMap options) {
    String selectorType = options.getString("selectorType");
    String selectorValue = null;
    if (options.hasKey("selectorValue")) {
      selectorValue = options.getString("selectorValue");
    }

    String actionType = options.getString("actionType");
    String actionValue = null;
    if (options.hasKey("actionValue")) {
      actionValue = options.getString("actionValue");
    }

    int delay = 0;
    if (options.hasKey("delay")) {
      delay = options.getInt("delay");
    }
    EventBus.getDefault().post(new TestActionEvent(selectorType, selectorValue, actionType, actionValue, delay));
  }

  @ReactMethod
  public void completed(final String stringifiedJson) {
    EventBus.getDefault().post(new TestCompletedEvent(stringifiedJson));
  }
}

File: apps/expo-go/android/expoview/src/main/java/versioned/host/exp/exponent/modules/test/ExponentTestNativeModule.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package versioned.host.exp.exponent.modules.test

import com.facebook.react.bridge.*
import de.greenrobot.event.EventBus
import host.exp.exponent.test.TestResolvePromiseEvent
import host.exp.exponent.generated.ExponentBuildConstants
import host.exp.exponent.kernel.KernelConfig
import host.exp.exponent.test.TestActionEvent
import host.exp.exponent.test.TestCompletedEvent
import org.json.JSONObject

class ExponentTestNativeModule(reactContext: ReactApplicationContext?) :
  ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String {
    return "ExponentTest"
  }

  private var currentId = 0
  private val idToPromiseMap = mutableMapOf<Int, Promise>()

  private fun getPromiseId(promise: Promise): Int {
    val id = currentId++
    idToPromiseMap[id] = promise
    return id
  }

  fun onEvent(event: TestResolvePromiseEvent) {
    if (idToPromiseMap.containsKey(event.id)) {
      idToPromiseMap[event.id]!!.resolve(true)
      idToPromiseMap.remove(event.id)
    }
  }

  override fun getConstants(): Map<String, Any> {
    return mapOf(
      "isInCI" to try {
        val config = JSONObject(ExponentBuildConstants.TEST_CONFIG)
        config.has("isInCI")
      } catch (e: Throwable) {
        false
      }
    )
  }

  @ReactMethod
  fun action(options: ReadableMap, promise: Promise) {
    if (!KernelConfig.IS_TEST) {
      promise.resolve(true)
    }

    val selectorType = options.getString("selectorType")
    val selectorValue = if (options.hasKey("selectorValue")) {
      options.getString("selectorValue")
    } else {
      null
    }

    val actionType = options.getString("actionType")
    val actionValue = if (options.hasKey("actionValue")) {
      options.getString("actionValue")
    } else {
      null
    }

    val timeout = if (options.hasKey("timeout")) {
      options.getInt("timeout")
    } else {
      0
    }

    val delay = if (options.hasKey("delay")) {
      options.getInt("delay")
    } else {
      0
    }

    EventBus.getDefault().post(
      TestActionEvent(
        getPromiseId(promise),
        selectorType,
        selectorValue,
        actionType,
        actionValue,
        delay,
        timeout
      )
    )
  }

  @ReactMethod
  fun completed(stringifiedJson: String, promise: Promise) {
    if (!KernelConfig.IS_TEST) {
      promise.resolve(true)
    }
    EventBus.getDefault().post(TestCompletedEvent(getPromiseId(promise), stringifiedJson))
  }

  companion object {
    private val TAG = ExponentTestNativeModule::class.java.simpleName
  }

  init {
    EventBus.getDefault().register(this)
  }
}
